### DISCUSSION TITLE
pattern_global_data

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/labview_book#22] [Announcements]
# 全局变量和功能全局变量

https://lv.qizhen.xyz/pattern_global_data

**Comment by Yang-Bo-Wen** [comment]
Event类实例的传递方式总是传引用嘛？

**Reply by ruanqizhen** [reply]
是的。虽然LabVIEW中大多数数据类型是值传递的，但是一些常常需要在多线程之间通信的数据类型（比如事件、队列等）还是被设计为了传引用，以方便不同线程之间的数据传递。

**Reply by SmallPeterZhu** [reply]
老师，我是做测试的，现在有点迷茫不知道怎么进步了。。这边用的上位机软件配合stm32开发板做伺服和变频器的测试，感觉很迷茫，老师有什么可指点的吗

**Reply by ruanqizhen** [reply]
> 老师，我是做测试的，现在有点迷茫不知道怎么进步了。。这边用的上位机软件配合stm32开发板做伺服和变频器的测试，感觉很迷茫，老师有什么可指点的吗

我不知道你是指项目上还是职业发展上的迷茫，不过在这两方面我可能都帮不少什么忙。如果有编程问题，我可能还帮得上，这是我擅长的领域。但是涉及到具体的应用，或者是更高的个人进步方面，我都不是专家，甚至也可以说水平也挺差的。我的建议是你可以跟身边的人讨论讨论，尤其是那些自己做的不错又了解你的人，问问他们的意见可能会最有帮助。

**Reply by SmallPeterZhu** [reply]
感谢您的回复，我现在做的测试，感觉对PCBA的功能测试不够全面！然后又不知道怎么进步，主要没有什么参考对象，不知道去如何修改我原来的东西！不知道老师您对变频器PCBA的测试有没有接触过

**Reply by ruanqizhen** [reply]
我也没有接触过。不过现在互联网上的技术资料非常的丰富，只要不是太过高深的技术，都有一大堆的文档，视频可以参考。首先Google一下（如果可以，搜索技术文章尽量用Google，搜索质量比国内几个搜索引擎好得多）。

**Comment by sixrui** [comment]
图 3.51 找不到呢

**Reply by ruanqizhen** [reply]
多谢提醒！在更新过程中有一些图片编号错乱了。我已经修改了这里，刷新浏览器就可以看到修改后的内容了。

**Comment by chenkekkk** [comment]
老师您好，提到这个功能全局变量，我看网上大多数都是处理单个数值的例子，有没有针对大数组的例子呢，还有就是簇的例子，如何做到在不同的vi中修改簇的某一个元素，还能原封不动将其他元素捆绑回去呢

**Reply by ruanqizhen** [reply]
举例一般总是会使用简单的例子。但在以前的实际项目中，单个数据一般用不着复杂的机制进行处理，功能全局变量更多的还是用于簇这种复杂数据的处理。只要合理的使用捆绑松绑节点配合移位寄存器对簇的元素进行处理，就可以达到你描述的功能。具体细节可以参考本书对于[簇数据](https://lv.qizhen.xyz/data_array#%E7%B0%87)的介绍。
目前，如果需要对于复杂数据，比如簇或数组进行处理，更推荐使用面向对象的编程方法，而不是使用全局变量，具体可以参考[面向对象编程](https://lv.qizhen.xyz/oop__)一章。

**Comment by ruanqizhen** [comment]
之前博客相关内容下的留言：

abby
2009年08月12日 19:56 编辑
今天用labview自带的全局变量和自己写的function gloabl做了俩个程序，验证了下运行效率和内存消耗，似乎后者的消耗是前者的几百倍，大出乎我的意料，不知道是不是我的function gloabl有问题，想让您帮着看一下。如有时间，请回复我haofang021@126.com.我将程序发给您，谢谢！

回复
Qizhen
2009年01月8日 21:20 编辑
function global 中的数据在停止VI运行时不会丢失，但关闭VI后就会失去。function global可以用在多个VI中。

回复
gufei
2009年01月7日 20:16 编辑
看来好像是哦：0

回复
gufei
2009年01月7日 20:15 编辑
即function global是不是只能在一个vi中使用？

回复
gufei
2009年01月7日 19:48 编辑
请问，功能性全局变量的生存周期是不是只能是本vi啊？本vi一停止，那这块数据内存就释放了啊。。。可否指教一下？如果我做菜单形式的程序，点击一个ITEM，动态调用一个子VI读入数据（使用功能全局变量），那这个子vi一关，那不就读不到这个数据了？在线等答案，先谢谢了啊：）

回复
Qizhen
2009年01月7日 10:17 编辑
我写的比较慢，等写完了会通知大家的。估计至少今年下半年才可能看到书吧。还没起名。

回复
gufei
2009年01月6日 09:35 编辑
请问高人，你的书什么时候才能买到啊？叫啥名？望眼欲穿啊

回复
lijuansophie
2008年07月17日 16:33 编辑
受教了，实在是高人！！

回复
静
2008年06月3日 13:21 编辑
很受用，谢谢！

回复
磊
2008年05月27日 14:45 编辑
先顶了再慢慢看，长知识

**Comment by robertlzj** [comment]
我理解，'功能全局变量'的本质是把数据、操作打包到单例的VI，成为原子操作。

**Comment by ruanqizhen** [comment]
文中提出了一个问题： 我们使用信号量保护了自己的数据，使得它们不会进入竞争状态。那么，谁来保护这个信号量呢？如果两个线程（运行在两个 CPU 上），在同一时钟周期里去锁同一个信号量，是不是就出现竞争状态了？

答案是：操作系统和 CPU 硬件。 信号量的获取操作在底层是由 CPU 的 “原子指令”（Atomic Instruction，如 Test-and-Set）来保证的。原子指令能确保在读取并修改内存标记的过程中，绝对不会被其他线程或 CPU 核心打断。LabVIEW 的信号量 VI 只是对操作系统这些底层能力的封装。
